### PR TITLE
Adds a working capabilities service.

### PR DESCRIPTION
**Pull request by LukeButters**
# Background

Adds a **working** capabilities service, adding a V2 of the capabilities service removing the v1 service.

The first service had a bug in which newtonsoft within halibut did not know how to deserialize a `IList` also the backwards comparability didn't work since tentacle made Halibut throw a different exception. The bug was not detected until the service was used since it only happened when halibut was involved. This PR uses the recently added tests infrastructure to verify this service works.

The changes are:

- Remove the old capabilities service.
- Adds a new capabilities service (v2).
    - The service returns a `List` not a `IList`.
- Updates the backwards compatibility decorator to work with both the standard halibut exception type as well as the exception returned by Tentacle when the service is unknown,
- Adds an integration tests with real tentacle to verify that added capabilities service works end to end.
- Adds an integration test with a real old Tentacle which does not have the capabilities service to verify the backwards compatibility works.

[SC-46171]
[SC-47868]

[Passing full chain build](https://build.octopushq.com/buildConfiguration/OctopusDeploy_OctopusServer_TeamCorePlatform_ChainFull/7808339?buildTab=dependencies&mode=timeline&type=snapshot&range=1684298434000-1684305159000) related [PR](https://github.com/OctopusDeploy/OctopusDeploy/pull/17972/files)

Also appears to work within Octopus.

## Backwards compatibility

Previously the AutfacServiceFactory was in octopus shared, [this](https://github.com/OctopusDeploy/OctopusShared/commits/e1afcc9a7357b3c6c895fa97a1bb0a79f84444e3/source/Octopus.Shared/Communications/AutofacServiceFactory.cs) looks like the complete history of the file. It was removed in this https://github.com/OctopusDeploy/OctopusShared/pull/149

At that point the exception would be whatever autofac happens to throw if it can not register a service specifically when:
```
var service = scope.ResolveNamed<object>(serviceName);
```
failed.

So that `BackwardsCompatibleCapabilitiesV2Decorator` always works with old tentacle. The code here has taken the conservative approach of concluding that if any exception is one with a stack trace within the  `CreateService` method (for any implementation) then the service can not be found.

It goes further and looks for other strings as well including "NotRegistered".

An example stack trace from version 3.0.0 tentacle would match and indeed when manually testing that version of tentacle works with the `BackwardsCompatibleCapabilitiesV2Decorator`.

```
Halibut.HalibutClientException: The requested service 'ICapabilitiesServiceV2 (System.Object)' has not been registered. To avoid this exception, either register a component to provide the service, check for service registration using IsRegistered(), or use the ResolveOptional() method to resolve an optional dependency.

Server exception: 
Autofac.Core.Registration.ComponentNotRegisteredException: The requested service 'ICapabilitiesServiceV2 (System.Object)' has not been registered. To avoid this exception, either register a component to provide the service, check for service registration using IsRegistered(), or use the ResolveOptional() method to resolve an optional dependency.
   at Autofac.ResolutionExtensions.ResolveService(IComponentContext context, Service service, IEnumerable`1 parameters)
   at Autofac.ResolutionExtensions.ResolveNamed[TService](IComponentContext context, String serviceName, IEnumerable`1 parameters)
   at Octopus.Shared.Communications.AutofacServiceFactory.CreateService(String serviceName) in y:\work\refs\heads\release\source\Octopus.Shared\Communications\AutofacServiceFactory.cs:line 18
   at Halibut.ServiceModel.ServiceInvoker.Invoke(RequestMessage requestMessage) in y:\work\7ab39c94136bc5c6\source\Halibut\ServiceModel\ServiceInvoker.cs:line 21
   at Halibut.Transport.Protocol.MessageExchangeProtocol.InvokeAndWrapAnyExceptions(RequestMessage request, Func`2 incomingRequestProcessor) in y:\work\7ab39c94136bc5c6\source\Halibut\Transport\Protocol\MessageExchangeProtocol.cs:line 139
   at Halibut.ServiceModel.HalibutProxy.ThrowExceptionFromReceivedError(ServerError error, ILog logger)
   at Halibut.ServiceModel.HalibutProxy.EnsureNotError(ResponseMessage responseMessage)
   at Halibut.ServiceModel.HalibutProxy.Invoke(MethodInfo targetMethod, Object[] args)
   at generatedProxy_3.GetCapabilities()
   at Octopus.Tentacle.Contracts.Capabilities.BackwardsCompatibleCapabilitiesV2Decorator.GetCapabilities() in C:\Users\auser\Documents\octopus\OctopusTentacle\source\Octopus.Tentacle.Contracts\Capabilities\BackwardsCompatibleCapabilitiesV2Decorator.cs:line 21
```

# Other additions.

- This PR also adds the ability to run version 5 tentacles on windows in tests.-
- Fixes a few bugs in the polling tentacle builder.
# Results

Hopefully a capabilities service that works.

# How to review this PR

Quality :heavy_check_mark:

# Pre-requisites

- [ ] I have read [How we use GitHub Issues](https://github.com/OctopusDeploy/Issues/blob/master/docs/CONTRIBUTING.internal.md) for help deciding when and where it's appropriate to make an issue.
- [ ] I have considered informing or consulting the right people, according to the [ownership map](https://whimsical.com/ownership-map-NzbiD4HJyvhC9jNJNfS6TG).
- [ ] I have considered appropriate testing for my change.

### CODE REVIEW

File: source/Octopus.Tentacle.Tests.Integration/CapabilitiesServiceV2Test.cs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+using NUnit.Framework.Internal;
+using Octopus.Tentacle.Contracts;
+using Octopus.Tentacle.Contracts.Legacy;
+using Octopus.Tentacle.Tests.Integration.Support;
+using Octopus.Tentacle.Tests.Integration.TentacleClient;
+using Octopus.Tentacle.Util;
+
+namespace Octopus.Tentacle.Tests.Integration
+{
+    public class CapabilitiesServiceV2Test
```
Comment (LukeButters) [inline source/Octopus.Tentacle.Tests.Integration/CapabilitiesServiceV2Test.cs:17]: The actually valuable tests :D

File: source/Octopus.Tentacle.Tests/Capabilities/CapabilitiesServiceV2Fixture.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    public class CapabilitiesServiceV2Fixture
+    {
+        [Test]
+        public void CapabilitiesAreReturned()
+        {
+            var capabilities = new CapabilitiesServiceV2()
+                .GetCapabilities()
+                .SupportedCapabilities;
+
+            capabilities.Should().Contain("ScriptService");
```
Comment (nathanwoctopusdeploy) [inline source/Octopus.Tentacle.Tests/Capabilities/CapabilitiesServiceV2Fixture.cs:16]: Were we not going to return the interfaces e.g. `nameof(IScriptService)`.

This allows a caller to also use the contract names for discovery and reduces a few magic strings.

File: source/Octopus.Tentacle/Services/Capabilities/CapabilitiesServiceV2.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+using Octopus.Tentacle.Contracts.Capabilities;
+
+namespace Octopus.Tentacle.Services.Capabilities
+{
+    [Service]
+    public class CapabilitiesServiceV2 : ICapabilitiesServiceV2
+    {
+        public CapabilitiesResponseV2 GetCapabilities()
+        {
+            return new CapabilitiesResponseV2(new List<string>() {"ScriptService", "FileTransferService"});
```
Comment (nathanwoctopusdeploy) [inline source/Octopus.Tentacle/Services/Capabilities/CapabilitiesServiceV2.cs:11]: Were we not going to return the interfaces e.g. nameof(IScriptService).

This allows a caller to also use the contract names for discovery and reduces a few magic strings.

Comment (LukeButters) [reply source/Octopus.Tentacle/Services/Capabilities/CapabilitiesServiceV2.cs:11]: Oh I see that makes more sense.

**Review by nathanwoctopusdeploy — APPROVED** [review]
LGTM - just the naming of the capabilities

File: source/Octopus.Tentacle.Contracts/Capabilities/BackwardsCompatibleCapabilitiesV2Decorator.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+            }
+            catch (HalibutClientException e) when 
+                (
+                    ExceptionLooksLikeTheServiceWasNotFound(e))
+            {
+                return new CapabilitiesResponseV2(new List<string>() {nameof(IScriptService), nameof(IFileTransferService)});
+            }
+        }
+
+        private static bool ExceptionLooksLikeTheServiceWasNotFound(HalibutClientException e)
```
Comment (LukeButters) [inline source/Octopus.Tentacle.Contracts/Capabilities/BackwardsCompatibleCapabilitiesV2Decorator.cs:35]: @nathanwoctopusdeploy it is worth re-reviewing this section which adds support for older tentacles.

**Comment by nathanwoctopusdeploy** [comment]
[sc-46268]

File: source/Octopus.Tentacle.Tests.Integration/CapabilitiesServiceV2Test.cs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+using NUnit.Framework;
+using Octopus.Tentacle.Contracts.Legacy;
+using Octopus.Tentacle.Tests.Integration.Support;
+using Octopus.Tentacle.Tests.Integration.TentacleClient;
+
+namespace Octopus.Tentacle.Tests.Integration
+{
+    public class CapabilitiesServiceV2Test
+    {
+        [TestCase(null)] // Current version
```
Comment (nathanwoctopusdeploy) [inline source/Octopus.Tentacle.Tests.Integration/CapabilitiesServiceV2Test.cs:15]: `"latest"`

File: source/Octopus.Tentacle.Tests/Capabilities/CapabilitiesServiceV2Fixture.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    public class CapabilitiesServiceV2Fixture
+    {
+        [Test]
+        public void CapabilitiesAreReturned()
+        {
+            var capabilities = new CapabilitiesServiceV2()
+                .GetCapabilities()
+                .SupportedCapabilities;
+
+            capabilities.Should().Contain("IScriptService");
```
Comment (nathanwoctopusdeploy) [inline source/Octopus.Tentacle.Tests/Capabilities/CapabilitiesServiceV2Fixture.cs:16]: Ensure only these 2 are returned

**Review by nathanwoctopusdeploy — APPROVED** [review]
LGTM

File: source/Octopus.Tentacle.Tests/Capabilities/CapabilitiesServiceV2Fixture.cs (hunk 6 above)
Comment (LukeButters) [reply source/Octopus.Tentacle.Tests/Capabilities/CapabilitiesServiceV2Fixture.cs:16]: done

**Comment by nathanwoctopusdeploy** [comment]
Fixes: https://github.com/OctopusDeploy/Issues/issues/8229

**Pull request merged**